File: ToC_2026/8/question4/C/4.c
//2つの自然数 m と n を取り，m を n で割った余りを返す関数 mod
#include <stdio.h>
int iszero(int m) { return m == 0; }
int pi(int m) { if (m == 0) return 0; return m - 1; }

int lt(int m, int n) {
    if (iszero(n)) return 0;
    else if (iszero(m)) return 1;
    else return lt(pi(m), pi(n));
}

int cut_off_sub(int m, int n) {
    if (iszero(n)) return m;
    else return cut_off_sub(pi(m), pi(n));
}

int mod(int m, int n) {
    if (iszero(n)) {
        return -1; 
    }
    
    if (lt(m, n)) {
        return m;
    } else {
        return mod(cut_off_sub(m, n), n);
    }
}

int main() {
    int m, n;
    printf("Enter two numbers (m n): ");
    scanf("%d %d", &m, &n);
    int result = mod(m, n);
    if (result == -1) {
        printf("Error: Cannot divide by zero.\n");
    } else {
        printf("The remainder of %d divided by %d is %d\n", m, n, result);
    }
    return 0;
}

File: ToC_2026/8/question4/Haskell/4.hs
iszero :: Int -> Bool
iszero m = m == 0

pi_ :: Int -> Int
pi_ m = if m == 0 then 0 else m - 1

lt :: Int -> Int -> Bool
lt m n = if iszero n then False else if iszero m then True else lt (pi_ m) (pi_ n)

cut_off_sub :: Int -> Int -> Int
cut_off_sub m n = if iszero n then m else cut_off_sub (pi_ m) (pi_ n)

mod_ :: Int -> Int -> Int
mod_ m n = if iszero n
           then -1
           else if lt m n
                then m
                else mod_ (cut_off_sub m n) n

main :: IO ()
main = do
    putStrLn "Enter two numbers (m n): "
    [m, n] <- map read . words <$> getLine
    let result = mod_ m n
    if result == -1
        then putStrLn "Error: Cannot divide by zero."
        else putStrLn $ "The remainder of " ++ show m ++ " divided by " ++ show n ++ " is " ++ show result